analytics/lib/counts.py

```diff
@@ -433,6 +433,33 @@ def pull_function(
     return DataCollector(output_table, pull_function)
 
 
+def count_upload_space_used_by_realm_query(realm: Optional[Realm]) -> QueryFn:
+    if realm is None:
+        realm_clause: Composable = SQL("")
+    else:
+        realm_clause = SQL("zerver_attachment.realm_id = {} AND").format(Literal(realm.id))
+
+    # TODO: What about ArchivedAttachments?
+
+    return lambda kwargs: SQL(
+        """
+            INSERT INTO analytics_realmcount (realm_id, property, end_time, value)
+            SELECT
+                zerver_attachment.realm_id,
+                %(property)s,
+                %(time_end)s,
+                COALESCE(SUM(zerver_attachment.size), 0)
+            FROM
+                zerver_attachment
+            WHERE
+                {realm_clause}
+                zerver_attachment.create_time < %(time_end)s
+            GROUP BY
```

Comment:
This query walks the entire table each time right now. I think this is likely to be unfeasible in production? So this would have to be more fancy and just do `previous countstat` + `size of new attachments` instead. But then some inaccuracy drift from the scenario https://github.com/zulip/zulip/pull/29789/files#r1571601458 might occur if I'm thinking right about this.

Reply:
Ugh, I think this may be forced to walk the whole table? Because with the `previous countstat + size of new attachments` approach, this number can only increase over time and this mechanism cannot account for the fact the organization can delete some files to free up space.

Here's the query I had to do this, but I don't how this can be adapted to deal with possibility of deleted attachments freeing up space.
```
WITH LatestCounts AS (
    SELECT
        ar.realm_id,
        ar.end_time AS latest_end_time,
        ar.value AS latest_value
    FROM
        analytics_realmcount ar
    WHERE
        ar.property = 'used_upload_space::day'
        AND ar.end_time = (
            SELECT MAX(ar2.end_time)
            FROM analytics_realmcount ar2
            WHERE ar2.realm_id = ar.realm_id AND ar2.property = 'used_upload_space::day'
        )
),
AttachmentSum AS (
    SELECT
        za.realm_id,
        COALESCE(SUM(za.size), 0) AS new_size
    FROM
        zerver_attachment za
    LEFT JOIN LatestCounts lc ON za.realm_id = lc.realm_id
    WHERE
        za.create_time > COALESCE(lc.latest_end_time, '1970-01-01')
    GROUP BY
        za.realm_id
)
SELECT
    COALESCE(asum.realm_id, lc.realm_id) AS realm_id,
    'used_upload_space::day' AS property,
    NOW() AS end_time,
    COALESCE(lc.latest_value, 0) + COALESCE(asum.new_size, 0) AS value
FROM
    LatestCounts lc
FULL OUTER JOIN AttachmentSum asum ON lc.realm_id = asum.realm_id;
```

Reply:
I think we CAN do deltas correctly by subtracting the `ArchivedAttachment` entries, if we had dates for when they moved, which is available on `ArchiveTransaction` ... as long as nobody restores a transaction; and that could be patched by adding a `restored_timestamp` on `ArchiveTransaction` so we can add it back in on the right day.

But walking the table daily is not prohibitively expensive; it's like 100x smaller than Message, so likely walking its entirety will not be a more expensive operation than walking the last day of Message, and I don't really expect it to be a scaling consideration, so we should probably do the simple option.

So my temptation would be to just walk the `Attachment` table, maybe leaving a comment noting that possibility for the future, and maybe do a quick extension of the `ArchiveTransaction` model to store the timestamp that a restoration occurred, as that just seems like good general debugging field to have. 

@alexmv thoughts?

Reply:
That seems fine to me.

Reply:
> I think we CAN do deltas correctly by subtracting the `ArchivedAttachment` entries, if we had dates for when they moved, which is available on `ArchiveTransaction` ... as long as nobody restores a transaction; and that could be patched by adding a `restored_timestamp` on `ArchiveTransaction` so we can add it back in on the right day.

Hmm, this isn't enough. You can also delete attachments via the API `DELETE /attachments/<int:attachment_id>`, which doesn't create ArchivedAttachments. We'd need to change it do that, or maybe add `RealmAuditLog` records for that actions so that we use it for subtractions. The final query accounting for all this probably going to end up pretty gnarly. I'll leave a note like this in the code here:
```
    # Note: This query currently has to go through the entire table,
    # summing all the sizes of attachments for every realm. This can be improved
    # by having a query which looks at the latest CountStat for each realm,
    # and sums it with only the new attachments.
    # There'd be additional complexity added by the fact that attachments can
    # also be deleted. Partially this can be accounted for by subtracting
    # ArchivedAttachment sizes, but there's still the issue of attachments
    # which can be directly deleted via the API.
```

Reply:
Sounds good, thanks for recording that bit of understanding about our algorithm options!